csrc/mlkem_spi.cpp

```
@@ -0,0 +1,117 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#include "buffer.h"
+#include "env.h"
+#include "generated-headers.h"
+#include "util.h"
+#include "auto_free.h"
+#include <openssl/evp.h>
+#include <openssl/err.h>
+
+using namespace AmazonCorrettoCryptoProvider;
+
+// ML-KEM constants
+#define MLKEM_SHARED_SECRET_SIZE 32
+#define MLKEM_512_CIPHERTEXT_SIZE  768
+#define MLKEM_768_CIPHERTEXT_SIZE  1088  
+#define MLKEM_1024_CIPHERTEXT_SIZE 1568
+
+static int ciphertextLengthToParameterSet(size_t ciphertext_len) {
+    switch (ciphertext_len) {
+        case MLKEM_512_CIPHERTEXT_SIZE:  return 512;   
+        case MLKEM_768_CIPHERTEXT_SIZE:  return 768;   
+        case MLKEM_1024_CIPHERTEXT_SIZE: return 1024;  
+        default:   return -1;
+    }
+}
+
+JNIEXPORT jint JNICALL Java_com_amazon_corretto_crypto_provider_KemUtils_nativeGetParameterSet(
+    JNIEnv* pEnv, jclass, jlong evpKeyPtr)
+{
+    try {
+        raii_env env(pEnv);
+        if (unlikely(!evpKeyPtr)) {
+            throw_java_ex(EX_NPE, "Null key pointer");
+        }
+        
+        EVP_PKEY* key = reinterpret_cast<EVP_PKEY*>(evpKeyPtr);
+        EVP_PKEY_CTX_auto ctx = EVP_PKEY_CTX_auto::from(EVP_PKEY_CTX_new(key, NULL));
+        if (!ctx.isInitialized()) {
+            throw_java_ex(EX_RUNTIME_CRYPTO, "Failed to create EVP context");
+        }
+        
+        size_t ciphertext_len, shared_secret_len;
+        CHECK_OPENSSL(EVP_PKEY_encapsulate(ctx, NULL, &ciphertext_len, NULL, &shared_secret_len));
+        
+        int paramSet = ciphertextLengthToParameterSet(ciphertext_len);
+        if (paramSet == -1) {
+            throw_java_ex(EX_RUNTIME_CRYPTO, "Unknown ML-KEM parameter set");
+        }
```

This check is likely unnecessary. You are checking if `paramSet == -1` here in native code, and if it is then you throw an exception. However, in the corresponding Java code you are doing the opposite conversion, catching the exception and converting it to -1. You should throw the exception, but probably from Java code.

---

Changed in next commit to throw exception on Java side.